frame/support/procedural/src/pallet/parse/storage.rs

```
@@ -695,21 +697,111 @@ impl StorageDef {
+												"Invalid pallet::storage, unexpected number of \
+												path segments for the generics in ResultQuery, \
+												expected a path with at least 2 segments, found {}",
+												err_variant.len(),
+											);
+											return Err(syn::Error::new(err_variant.span(), msg))
+										}
+										let mut error = err_variant.clone();
+										let err_variant = error
+											.pop()
+											.expect("Checked to have at least 2; qed")
+											.into_value()
+											.ident;
+
+										// Necessary here to eliminate the last double colon
+										let last = error
+											.pop()
+											.expect("Checked to have at least 2; qed")
+											.into_value();
+										error.push_value(last);
+
+										Ok(Some(QueryKind::ResultQuery(
+											syn::Path {
+												leading_colon: leading_colon.clone(),
+												segments: error,
+											},
+											err_variant,
+										)))
+									},
+									gen_arg => {
+										let msg = format!(
+											"Invalid pallet::storage, unexpected generic argument \
+											kind, expected a type path to a `PalletError` enum \
+											variant, found `{}`",
+											gen_arg.to_token_stream().to_string(),
+										);
+										Err(syn::Error::new(gen_arg.span(), msg))
+									},
+								}
+							},
+							args => {
+								let msg = format!(
+									"Invalid pallet::storage, unexpected generic args for \
+									ResultQuery, expected angle-bracketed arguments, found `{}`",
+									args.to_token_stream().to_string()
+								);
+								Err(syn::Error::new(args.span(), msg))
+							},
+						}
+					},
+					Type::Path(path)
+						if path.path.segments.last().map_or(false, |s| s.ident == "ValueQuery") =>
+						Ok(Some(QueryKind::ValueQuery)),
+					_ => Ok(None),
+				}
 			})
-			.unwrap_or(Some(QueryKind::OptionQuery)); // This value must match the default generic.
+			.transpose()?
+			.unwrap_or(Some(QueryKind::OptionQuery));
```

Not part of your PR, but kinda strange that this is the behavior: why if we fail to parse the query kind, we silently fallback to option?

---

Or is this the case where we don't have any query type provided?

---

Because `OptionQuery` is the default.